# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!


In [1]:
#Import your libraries
from sklearn.model_selection import train_test_split
import numpy as np
import pandas as pd


# Challenge 1 -  Explore the Internal Dataset

In this lab, we will start off by working with the wine dataset in scikit-learn. We will select the wine dataset and use a clustering algorithm to learn more about the functionalities of this library. 

We start off by loading the dataset using the `load_wine` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_wine.html)). In the cell below, we will import the function from scikit-learn.

In [2]:
from sklearn.datasets import load_wine

In the cell below, use the `load_wine` function and assign the wine dataset to a variable called `wine`.

In [3]:
# Your code here:
wine = load_wine()

In the next step, list the keys of the variable `wine` to examine its contents. Note that the `load_wine` function does not return dataframes. It returns you a Python dictionary.

In [4]:
# Your code here:
wine.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

Next, list the feature names. These are the different characteristics of the wine. 

In [5]:
# Your code here:
wine.feature_names

['alcohol',
 'malic_acid',
 'ash',
 'alcalinity_of_ash',
 'magnesium',
 'total_phenols',
 'flavanoids',
 'nonflavanoid_phenols',
 'proanthocyanins',
 'color_intensity',
 'hue',
 'od280/od315_of_diluted_wines',
 'proline']

Print the description of the dataset in the cell below using the `DESCR` attribute of the `wine` variable.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the print function.*

In [6]:
# Your code here:
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

#### From the description, we see that all columns are numeric. We also know that there is no missing data 

Let's plot the alcohol content histogram. Recall that we are working with a numpy array and will need to use a matplotlib function to produce a histogram. 

In [7]:
# imports
import matplotlib.pyplot as plt

In [8]:
# Your code here:
wine['data'][0]

array([1.423e+01, 1.710e+00, 2.430e+00, 1.560e+01, 1.270e+02, 2.800e+00,
       3.060e+00, 2.800e-01, 2.290e+00, 5.640e+00, 1.040e+00, 3.920e+00,
       1.065e+03])

# Challenge 2 - Clustering the Internal Dataset

In this portion of the lab, we will cluster the data to find common traits between the different wines. We will use the k-means clustering algorithm to achieve this goal.

#### We start by importing k-means from scikit-learn and then proceed to create 4 clusters.

In [9]:
# imports
from sklearn.cluster import KMeans

In [10]:
# Your code here:
knncluster = KMeans(n_clusters=4).fit(wine['data'])

#### Print the cluster labels.

In [11]:
# Your code here:
knncluster.labels_

array([0, 0, 2, 2, 3, 2, 2, 2, 0, 0, 2, 2, 2, 0, 2, 2, 2, 0, 2, 3, 3, 3,
       0, 0, 3, 3, 2, 2, 0, 0, 2, 2, 0, 2, 0, 0, 0, 0, 0, 3, 3, 0, 0, 3,
       0, 0, 0, 0, 0, 2, 0, 2, 2, 2, 0, 0, 0, 2, 2, 1, 3, 1, 3, 1, 1, 3,
       1, 1, 3, 3, 0, 1, 1, 0, 0, 1, 1, 1, 3, 1, 1, 3, 3, 1, 1, 1, 1, 1,
       3, 3, 1, 1, 1, 1, 1, 0, 3, 1, 3, 1, 3, 1, 1, 1, 3, 1, 1, 1, 1, 3,
       1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 3, 3, 1,
       1, 3, 3, 3, 3, 1, 3, 3, 3, 3, 1, 1, 3, 3, 1, 3, 3, 1, 1, 1, 1, 3,
       3, 3, 1, 0, 3, 3, 1, 3, 1, 3, 3, 1, 3, 3, 3, 3, 1, 1, 3, 3, 3, 3,
       3, 1])

#### Compute the size of each cluster. This can be done by counting the number of occurrences of each unique label in the list above.

Which is the largest cluster of the 4?

In [12]:
# Your code here:
unique, counter = np.unique(knncluster.labels_, return_counts=True)


In [13]:
# Your answer here:
dict(zip(unique,counter)) #0 is the largest one

{0: 32, 1: 66, 2: 23, 3: 57}

#### Inspect the shape of `wine['data']`

In [14]:
# Your code here:
wine['data'].shape

(178, 13)

#### Inspect the first 5 records in `wine['data']`

In [15]:
# Your code here:
wine['data'][:5]

array([[1.423e+01, 1.710e+00, 2.430e+00, 1.560e+01, 1.270e+02, 2.800e+00,
        3.060e+00, 2.800e-01, 2.290e+00, 5.640e+00, 1.040e+00, 3.920e+00,
        1.065e+03],
       [1.320e+01, 1.780e+00, 2.140e+00, 1.120e+01, 1.000e+02, 2.650e+00,
        2.760e+00, 2.600e-01, 1.280e+00, 4.380e+00, 1.050e+00, 3.400e+00,
        1.050e+03],
       [1.316e+01, 2.360e+00, 2.670e+00, 1.860e+01, 1.010e+02, 2.800e+00,
        3.240e+00, 3.000e-01, 2.810e+00, 5.680e+00, 1.030e+00, 3.170e+00,
        1.185e+03],
       [1.437e+01, 1.950e+00, 2.500e+00, 1.680e+01, 1.130e+02, 3.850e+00,
        3.490e+00, 2.400e-01, 2.180e+00, 7.800e+00, 8.600e-01, 3.450e+00,
        1.480e+03],
       [1.324e+01, 2.590e+00, 2.870e+00, 2.100e+01, 1.180e+02, 2.800e+00,
        2.690e+00, 3.900e-01, 1.820e+00, 4.320e+00, 1.040e+00, 2.930e+00,
        7.350e+02]])

You now know the data object is a 2-dimensional array in which there are 178 rows and 13 columns. Each row is a data record and each column is a feature.

#### What is the average ash content for each cluster? 

*Hints:* 

* *Ash* is the 3rd column.

* The data object is not a Pandas dataframe so you can't apply `pandas.DataFrame.groupby`. Instead, you can use `np.average`.

In [16]:
# Your code here

np.average(list(zip(knncluster.labels_,wine['data'][:,2])))

1.9782022471910112

array([14.23,  1.71,  2.43, ...,  3.  ,  3.  ,  0.  ])

# Challenge 3 - Load and Explore an External Dataset

We will now load an external dataset using Pandas and use scikit learn to explore the data. In this portion of the lab, we will use a [patient dataset from Kaggle](https://www.kaggle.com/miles99/patient-admission-dataset-for-learning-data-mining). 

In [17]:
patients = pd.read_csv('../patient-admission-dataset-for-learning-data-mining.csv')

In the next cell, print the first five rows of the data using the `head()` function.

In [18]:
# Your code here:
patients.head()

id         patient_name               patient_email  doctor_phone  \
0   1  Celestyna Dillimore      cdillimore0@dion.ne.jp  674-914-1212   
1   2        Meta Michieli          mmichieli1@loc.gov  172-580-3586   
2   3        Cordie Sancto      csancto2@cafepress.com  794-222-5085   
3   4     Josh De Ambrosis           jde3@amazon.co.jp  856-540-5195   
4   5    Delinda Alfonsini  dalfonsini4@opensource.org  938-978-1131   

  patient_gender patient_dob  patient_diabetic  patient_allergic  \
0         Female  10/18/2018             False              True   
1         Female    2/8/2018             False              True   
2         Female   10/9/2018              True              True   
3           Male   9/10/2018              True              True   
4         Female   2/26/2018             False              True   

   patient_weight_kg  patient_height_sm  patient_nhs_number  \
0                 59                176        8.200152e+09   
1                 77                186                 NaN   
2                 90                177        6.145594e+09   
3                 70                150                 NaN   
4                 82                140        4.804758e+08   

           doctor_name appointment_date  patient_show  is_regular_visit  \
0      Sarena Waliszek         5/1/2018          True              True   
1       Farris Robinet        12/7/2017          True              True   
2       Kaspar Spitaro        10/5/2018         False             False   
3       Rafferty Fowls       10/21/2018         False              True   
4  Glenna MacNeachtain       11/15/2018         False             False   

      prescribed_medicines diagnosis  
0  triamcinolone acetonide      I669  
1                      NaN       NaN  
2                      NaN       NaN  
3                      NaN       NaN  
4                      NaN       NaN

Next, print the column types and check which columns have been misclassified by pandas.

In [19]:
# Your code here:
patients.dtypes

id                        int64
patient_name             object
patient_email            object
doctor_phone             object
patient_gender           object
patient_dob              object
patient_diabetic           bool
patient_allergic           bool
patient_weight_kg         int64
patient_height_sm         int64
patient_nhs_number      float64
doctor_name              object
appointment_date         object
patient_show               bool
is_regular_visit           bool
prescribed_medicines     object
diagnosis                object
dtype: object

#### We can see that none of the date columns have been correctly classified. Also, some columns contain qualitative data that can be dropped.

First, transform the `patient_dob` and `appointment_date` columns to datetime using the `pd.to_datetime` function.

In [20]:
# Your code here:
patients['patient_dob'] = pd.to_datetime(patients['patient_dob'])
patients['appointment_date']= pd.to_datetime(patients['appointment_date'])

Next, drop the `id`, `patient_name`, `patient_email`, `patient_nhs_number`, and `doctor_phone` columns. These are not quantitative columns and will not contribute to our analysis.

In [21]:
# Your code here:
patients.drop(columns=['id', 'patient_name', 'patient_email', 'patient_nhs_number','doctor_phone'], inplace=True)

Now we work on the missing data. Most ML algorithms will not perform as intended if there are missing data.

In the cell below, count how many rows contain missing data in each column. You should see three columns contain missing data:

* `doctor_name`: 58 missing data
* `prescribed_medicines`: 488 missing data
* `diagnosis`: 488 missing data

In [22]:
# Your code here:
patients.isnull().sum()

patient_gender            0
patient_dob               0
patient_diabetic          0
patient_allergic          0
patient_weight_kg         0
patient_height_sm         0
doctor_name              58
appointment_date          0
patient_show              0
is_regular_visit          0
prescribed_medicines    488
diagnosis               488
dtype: int64

The main issues are found in the `prescribed_medicines` and `diagnosis` columns. Can we simply drop these rows?

The answer is not yet. Because when there are missing data in these columns, it doesn't mean the data records are broken. Instead, it means no medication was prescribed and no diagnosis was recorded. Therefore, once we fill in the missing data these columns will be fine. But we'll revisit these columns and decide whether we will eventually drop them when we look at how many unique values are there in these categorical columns.  

For the `prescribed_medicines` column, fill the missing values with the value `no prescription`. For the `diagnosis` column, fill the missing values with `no diagnosis`.

*Hint: Use [`pandas.DataFrame.fillna`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.fillna.html).*

In [23]:
# Your code here:
patients['prescribed_medicines'].fillna(value='no prescription', inplace=True)

patients['diagnosis'].fillna(value='no diagnosis', inplace=True)


How about `doctor_name`? Since a doctor visit without a doctor name might not be meaningful, we will drop these rows.

In [24]:
# Your code here:
patients.dropna(inplace=True)

#### Another step in preprocessing that can be performed by scikit-learn is label encoding. 

We have 4 columns that are of `bool` type. We would like to convert them to an integer column containing either zero or one. We can do this using [scikit-learn's label encoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html).

In the cell below, import the label encoder and encode the 4 boolean columns (*patient_diabetic*, *patient_allergic*, *patient_show*, *is_regular_visit*) with `0` and `1`. 

In [25]:
# Your code here:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

le.fit(patients['patient_diabetic'])
patients['patient_diabetic'] = le.transform(patients['patient_diabetic'])

le.fit(patients['patient_allergic'])
patients['patient_allergic'] = le.transform(patients['patient_allergic'])

le.fit(patients['patient_show'])
patients['patient_show'] = le.transform(patients['patient_show'])

le.fit(patients['is_regular_visit'])
patients['is_regular_visit'] = le.transform(patients['is_regular_visit'])

In [26]:
patients.head()

patient_gender patient_dob  patient_diabetic  patient_allergic  \
0         Female  2018-10-18                 0                 1   
1         Female  2018-02-08                 0                 1   
2         Female  2018-10-09                 1                 1   
3           Male  2018-09-10                 1                 1   
4         Female  2018-02-26                 0                 1   

   patient_weight_kg  patient_height_sm          doctor_name appointment_date  \
0                 59                176      Sarena Waliszek       2018-05-01   
1                 77                186       Farris Robinet       2017-12-07   
2                 90                177       Kaspar Spitaro       2018-10-05   
3                 70                150       Rafferty Fowls       2018-10-21   
4                 82                140  Glenna MacNeachtain       2018-11-15   

   patient_show  is_regular_visit     prescribed_medicines     diagnosis  
0             1                 1  triamcinolone acetonide          I669  
1             1                 1          no prescription  no diagnosis  
2             0                 0          no prescription  no diagnosis  
3             0                 1          no prescription  no diagnosis  
4             0                 0          no prescription  no diagnosis

Print the data dtypes to confirm those four `bool` columns are converted to `int64`.

In [27]:
# Your code here:
patients.dtypes

patient_gender                  object
patient_dob             datetime64[ns]
patient_diabetic                 int64
patient_allergic                 int64
patient_weight_kg                int64
patient_height_sm                int64
doctor_name                     object
appointment_date        datetime64[ns]
patient_show                     int64
is_regular_visit                 int64
prescribed_medicines            object
diagnosis                       object
dtype: object

#### The last step is to handle the `object` data.

There are 4 `object` columns now: `patient_gender`, `doctor_name`, `prescribed_medicines`, and `diagnosis`. The gender columns

In the next cell, check the unique values of each of the `object` columns using `value_counts()`.

In [28]:
# Your code here:
display('patient_gender', patients['patient_gender'].value_counts(), 'doctor_name', patients['doctor_name'].value_counts(), 
        'prescribed_medicines', patients['prescribed_medicines'], 'diagnosis', patients['diagnosis'])

'patient_gender'

Female    476
Male      466
Name: patient_gender, dtype: int64

'doctor_name'

Sarena Waliszek           74
Cissy Markey              57
Farris Robinet            55
Flossy Canlin             55
Gabie Brafield            53
Glenna MacNeachtain       53
Jilly McGrill             53
Nerte Elmer               51
Oralie Swaffer            49
Kaspar Spitaro            48
Carly SperaJillie Katt    47
Wyn Grassett              47
Humfried Cartmel          47
Madelina Scurrell         46
Rafferty Fowls            45
Dulce McKerley            42
Rouvin McKinstry          42
Ryley Swallow             40
Hermina Domeny            38
Name: doctor_name, dtype: int64

'prescribed_medicines'

0                                triamcinolone acetonide
1                                        no prescription
2                                        no prescription
3                                        no prescription
4                                        no prescription
                             ...                        
994                                      no prescription
996                                      no prescription
997                                      no prescription
998                                      IRON SUPPLEMENT
999    aluminum hydroxide, magnesium carbonate, sodiu...
Name: prescribed_medicines, Length: 942, dtype: object

'diagnosis'

0              I669
1      no diagnosis
2      no diagnosis
3      no diagnosis
4      no diagnosis
           ...     
994    no diagnosis
996    no diagnosis
997    no diagnosis
998           S6610
999          I70429
Name: diagnosis, Length: 942, dtype: object

#### The number of unique values is large for all three columns except `patient_gender`. We will handle these columns differently.

For `diagnosis`, there are too many unique values which will make ML difficult. However, we can re-encode the values to either with or without diagnosis. Remember at an earlier step we filled in the missing values of this column with *no diagnosis*? We can re-encode *no diagnosis* to `0` and all other values to `1`. In this way we can tremendously simply this column.

For `prescribed_medicines`, we can drop this column because it is perfectly correlated with `diagnosis`. Whenever there is no diagnosis, there is no prescribed medicine. So we don't need to keep this duplicated data.

How about `doctor_name`? There are not excessive unique values but still quite many (19). We may either drop or keep it but keeping it will make the analysis more complicated. So due to the length of this lab let's drop it.

How about `gender`? This one is easy. Just like re-encoding the boolean values, we can re-encode gender to `0` and `1` because there are only 2 unique values.

In the next cells, do the following:

1. Create a new column called `diagnosis_int` that has `0` and `1` based on the values in `diagnosis`.

1. Create a new column called `patient_gender_int` that has `0` and `1` based on the values in `patient_gender`.

1. Drop the following columns: `doctor_name`, `diagnosis`, `prescribed_medicines`, and `patient_gender`.

In [29]:
# Your code here:
le.fit(patients['patient_gender'])
patients['patient_gender_int']= le.transform(patients['patient_gender'])

patients['diagnosis_int'] = patients['diagnosis'].where(patients['diagnosis']=='no diagnosis', other= 1 )
patients['diagnosis_int'].where(patients['diagnosis_int']!='no diagnosis', other= 0,inplace=True)

patients.drop(columns=['doctor_name', 'diagnosis', 'prescribed_medicines', 'patient_gender'],inplace=True)

Let's look at the head again to ensure the re-encoding and dropping are successful:

In [30]:
# Your code here:
patients.head()

patient_dob  patient_diabetic  patient_allergic  patient_weight_kg  \
0  2018-10-18                 0                 1                 59   
1  2018-02-08                 0                 1                 77   
2  2018-10-09                 1                 1                 90   
3  2018-09-10                 1                 1                 70   
4  2018-02-26                 0                 1                 82   

   patient_height_sm appointment_date  patient_show  is_regular_visit  \
0                176       2018-05-01             1                 1   
1                186       2017-12-07             1                 1   
2                177       2018-10-05             0                 0   
3                150       2018-10-21             0                 1   
4                140       2018-11-15             0                 0   

   patient_gender_int diagnosis_int  
0                   0             1  
1                   0             0  
2                   0             0  
3                   1             0  
4                   0             0

An interesting observation is that all patients are no older than 2 years. However, their weights and heights indicate that they are adults. This cannot be true. Therefore, we can either trust the weight and height columns or the DOB column. Since there are other columns that indicate that these are adults (they have emails, some have diabetes) we will drop the `patient_dob` column. We will also drop the `appointment_date` column since it has too many unique values to transform to a dummy variable. Drop the two columns in the cell below.

In [31]:
# Your code here:
patients.drop(columns=['patient_dob','appointment_date'], inplace=True)

#### Our data is now ready for clustering. Let's use k-means again.

We start by initializing and fitting a model in the cell below. Call this model patients_cluster.

In [32]:
# Your code here:
from sklearn.cluster import KMeans

K_Mean = KMeans(n_clusters=4)
patients_cluster = K_Mean.fit(patients)


Attach the labels to the dataframe. Do this by accessing the `labels_` in the `patients_cluster` model and assign them to a new column in `patients` that you will call `labels`.

In [33]:
# Your code here:
patients['labels'] = patients_cluster.labels_

Now using a `groupby`, find the mean of every variable in `patients` and group by the `labels` column. This summary will allow us to see how the patients differ between the clusters. Your output should look similar to the image below.

![groupby mean](../groupby-mean.png)

Additionally, add a comment to describe which columns have the largest difference between clusters.

In [34]:
# Your code here:
patients.groupby(by='labels').feimean()

patient_diabetic  patient_allergic  patient_weight_kg  \
labels                                                          
0               0.506494          0.493506          65.432900   
1               0.552511          0.525114          97.557078   
2               0.492063          0.488095          94.063492   
3               0.508333          0.533333          62.900000   

        patient_height_sm  patient_show  is_regular_visit  patient_gender_int  
labels                                                                         
0              152.354978      0.510823          0.558442            0.554113  
1              153.319635      0.488584          0.502283            0.520548  
2              178.686508      0.503968          0.531746            0.456349  
3              176.954167      0.475000          0.495833            0.454167

In [35]:
# Your comment here:
""" it looks like the largest difference in height is the cluster 2"""


# Bonus Challenge: Visualize K-Means Clusters

How did k-means cluster the data? You can obtain an intuitive view with a scatter plot. Generate a 2-d cluster plot below using `matplotlib`. You need to choose 2 of the features from your cleaned and transformed dataset, and use color to represent the cluster label generated from k-means.

If the scatter plot does not make any sense to you, it means the features you chose to visualize are not the right ones. You should be able to see 4 clear clusters with different colors in your visualization that suggests how k-means had clustered your data.

![Cluster Visualization](../clusters.png)

In [36]:
# Your code here:


Additionally, you can visualize the clusters in 3-D scatter plot. Give it a try below.

In [37]:
# Your code here:
